File: web/src/schemas/llm.ts
// Named LLM + live (voice) configurations: the install-wide lists, the active id,
// and the probe result both Test buttons return.
import { z } from 'zod'
import { SecretRef, SharedKey } from './primitives.ts'
import { CatalogModel } from './system.ts'

// Optional provider-library state for a config type (llm_configs.deps_status).
export const DepsStatus = z.object({
  ok: z.boolean(),
  extra: z.string(),
  install: z.string(),
})
export type DepsStatus = z.infer<typeof DepsStatus>

// Which key an actual call would send (llm_configs.key_source docstring).
export const KeySource = z.enum(['secret', 'shared', 'not_needed', 'none', 'subscription', 'cli_login'])
export type KeySource = z.infer<typeof KeySource>

export const LlmConfig = z.object({
  id: z.string(),
  name: z.string(),
  type: z.string(),
  model: z.string(),
  base_url: z.string(),
  host: z.string(),
  options: z.record(z.string(), z.unknown()),
  secret_id: z.string(),
  secret: SecretRef.nullable(),
  secret_missing: z.boolean(),
  key_source: KeySource,
  images: z.boolean(),
  deps: DepsStatus,
  shared_key: SharedKey,
  active: z.boolean(),
  // Added only for type === 'openai_subscription' (live ChatGPT sign-in state).
  signed_in: z.boolean().optional(),
})
export type LlmConfig = z.infer<typeof LlmConfig>

// Whichever of AG2ASSISTANT_LLM_PROVIDER / AG2ASSISTANT_MODEL is set, else null.
export const LlmEnvOverride = z.object({
  provider: z.string().optional(),
  model: z.string().optional(),
})
export type LlmEnvOverride = z.infer<typeof LlmEnvOverride>

export const LlmConfigList = z.object({
  configs: z.array(LlmConfig),
  active: z.string().nullable(),
  env_override: LlmEnvOverride.nullable(),
  provider_deps: z.record(z.string(), DepsStatus),
})
export type LlmConfigList = z.infer<typeof LlmConfigList>

// GET /api/llm-configs/models — the same {models, current, reason} envelope the ACP
// route uses, with provider_catalog.py's own reasons (lib/modelSuggest.ts REASON)
// instead of the coding-agent ones. `current` is always '' here: a provider names no
// model of its own. CatalogModel is shared with the ACP route, hence the import.
export const ProviderCatalog = z.object({
  models: z.array(CatalogModel),
  current: z.string(),
  reason: z.enum(['', 'unauthorized', 'unreachable', 'no_list_endpoint', 'not_probeable']),
})
export type ProviderCatalog = z.infer<typeof ProviderCatalog>

export const LlmConfigSaved = z.object({
  ok: z.literal(true),
  config: LlmConfig,
  active: z.string().nullable(),
})
export type LlmConfigSaved = z.infer<typeof LlmConfigSaved>

export const LiveKeySource = z.enum(['secret', 'shared', 'none'])
export type LiveKeySource = z.infer<typeof LiveKeySource>

export const LiveConfig = z.object({
  id: z.string(),
  name: z.string(),
  provider: z.string(),
  model: z.string(),
  voice: z.string(),
  secret_id: z.string(),
  secret: SecretRef.nullable(),
  secret_missing: z.boolean(),
  key_source: LiveKeySource,
  shared_key: SharedKey,
  active: z.boolean(),
})
export type LiveConfig = z.infer<typeof LiveConfig>

export const LiveProvider = z.object({
  name: z.string(),
  default_model: z.string(),
  default_voice: z.string(),
})
export type LiveProvider = z.infer<typeof LiveProvider>

export const LiveConfigList = z.object({
  configs: z.array(LiveConfig),
  active: z.string().nullable(),
  providers: z.array(LiveProvider),
})
export type LiveConfigList = z.infer<typeof LiveConfigList>

export const LiveConfigSaved = z.object({
  ok: z.literal(true),
  config: LiveConfig,
  active: z.string().nullable(),
})
export type LiveConfigSaved = z.infer<typeof LiveConfigSaved>

// Both Test buttons: a 502 failure is thrown by the transport, so only success lands here.
export const PingResult = z.object({
  ok: z.literal(true),
  reply: z.string(),
  latency_ms: z.number(),
})
export type PingResult = z.infer<typeof PingResult>

File: src/assistant/gateway/schemas/llm.py
"""Named LLM and live (voice) configurations. Mirrors web/src/schemas/llm.ts.

Both config views are shaped by the same rule: name the key a real call would
send, never the key itself. Hence the ``SecretRefOut``/``SharedKeyOut`` pair plus
a ``key_source`` token, which together let the UI say honestly why a
keyless-looking config still works.
"""

from typing import Any, Literal

from pydantic import BaseModel

from .primitives import SecretRefOut, SharedKeyOut
from .system import CatalogModelOut


class DepsStatusOut(BaseModel):
    """Optional provider-library state for a config type (llm_configs.deps_status)."""

    ok: bool
    extra: str
    install: str


# Which key an actual call would send (llm_configs.key_source).
KeySourceOut = Literal["secret", "shared", "not_needed", "none", "subscription", "cli_login"]


class LlmConfigOut(BaseModel):
    """One named LLM configuration as the API exposes it.

    ``signed_in`` is added only for ``type == "openai_subscription"`` — the live
    ChatGPT sign-in state the chip needs without a second fetch — so it carries a
    default and every route returning this model excludes unset fields.
    """

    id: str
    name: str
    type: str
    model: str
    base_url: str
    host: str
    options: dict[str, Any]
    # {tool id: options} — the provider-native tools switched on for this config.
    # Availability is per type; see assistant.builtin_tools.
    builtin_tools: dict[str, dict[str, Any]]
    secret_id: str
    secret: SecretRefOut | None
    secret_missing: bool
    key_source: KeySourceOut
    images: bool
    deps: DepsStatusOut
    shared_key: SharedKeyOut
    active: bool
    signed_in: bool | None = None


class LlmEnvOverrideOut(BaseModel):
    """The env pin banner payload: whichever of AG2ASSISTANT_LLM_PROVIDER /
    AG2ASSISTANT_MODEL is set. Each key is present only when its variable is,
    which is why both carry a default rather than being nullable-required."""

    provider: str | None = None
    model: str | None = None


class LlmConfigListResponse(BaseModel):
    """GET /api/llm-configs. ``provider_deps`` covers every known type, not just
    the configured ones — the "Add model" template grid reads it for types no
    config uses yet."""

    configs: list[LlmConfigOut]
    active: str | None
    env_override: LlmEnvOverrideOut | None
    provider_deps: dict[str, DepsStatusOut]
    # type -> the provider-native tool ids it offers, for every type (same reason
    # as provider_deps). Ids only — the labels are the web's.
    builtin_tools_by_type: dict[str, list[str]]


class LlmConfigSavedResponse(BaseModel):
    """POST /api/llm-configs and POST /api/llm-configs/{cid}."""

    ok: Literal[True]
    config: LlmConfigOut
    active: str | None


class ProviderCatalogResponse(BaseModel):
    """GET /api/llm-configs/models — a provider's model catalog in the same
    ``{models, current, reason}`` envelope the ACP route uses, with
    provider_catalog.py's own reasons. ``current`` is always empty here: a
    provider names no model of its own."""

    models: list[CatalogModelOut]
    current: str
    reason: Literal["", "unauthorized", "unreachable", "no_list_endpoint", "not_probeable"]


LiveKeySourceOut = Literal["secret", "shared", "none"]


class LiveConfigOut(BaseModel):
    """One named live (voice) configuration as the API exposes it."""

    id: str
    name: str
    provider: str
    model: str
    voice: str
    secret_id: str
    secret: SecretRefOut | None
    secret_missing: bool
    key_source: LiveKeySourceOut
    shared_key: SharedKeyOut
    active: bool


class LiveProviderOut(BaseModel):
    """A voice provider's defaults, which seed the add-form and the templates."""

    name: str
    default_model: str
    default_voice: str


class LiveConfigListResponse(BaseModel):
    """GET /api/live-configs."""

    configs: list[LiveConfigOut]
    active: str | None
    providers: list[LiveProviderOut]


class LiveConfigSavedResponse(BaseModel):
    """POST /api/live-configs and POST /api/live-configs/{cid}."""

    ok: Literal[True]
    config: LiveConfigOut
    active: str | None


class PingResultResponse(BaseModel):
    """Both Test buttons on success. A failure is a 502 built by hand, so the
    ``ok: false`` branch never rides this model."""

    ok: Literal[True]
    reply: str
    latency_ms: int
